### PR TITLE
AVRO-3380: Raise an exception if invalid number of bytes read

### PR DESCRIPTION
**Pull request by jjaakola-aiven**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira-3380](https://issues.apache.org/jira/browse/AVRO-3380/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-3380
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
* lang/py/avro/test/test_io.py::TestIncompatibleSchemaReading

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/py/avro/io.py [hunk 1]
```diff
@@ -663,63 +666,69 @@ def read_data(self, writers_schema: avro.schema.Schema, readers_schema: avro.sch
...
+                return self.read_enum(writers_schema, readers_schema, decoder)
+            if isinstance(writers_schema, avro.schema.ArraySchema) and isinstance(readers_schema, avro.schema.ArraySchema):
+                return self.read_array(writers_schema, readers_schema, decoder)
+            if isinstance(writers_schema, avro.schema.MapSchema) and isinstance(readers_schema, avro.schema.MapSchema):
+                return self.read_map(writers_schema, readers_schema, decoder)
+            if isinstance(writers_schema, avro.schema.RecordSchema) and isinstance(readers_schema, avro.schema.RecordSchema):
+                # .type in ["record", "error", "request"]:
+                return self.read_record(writers_schema, readers_schema, decoder)
+        except avro.errors.InvalidBytesRead as e:
+            decoder.reader.seek(0)
```
Comment (RyanSkraba) [inline lang/py/avro/io.py:728]: I'm not sure this is the right thing to do here --> I like the original message in `InvalidBytesRead` more than the one constructed by AvroTypeException.

Is there something nice that we can do in the avro.error hierarchy to make this useful?

File: lang/py/avro/io.py [hunk 2]
```diff
@@ -222,7 +222,10 @@ def read(self, n: int) -> bytes:
         """
         Read n bytes.
         """
-        return self.reader.read(n)
+        read_bytes = self.reader.read(n)
```
Comment (RyanSkraba) [inline lang/py/avro/io.py:225]: You mentioned that the original `avro-python3` code also checked that n is non-negative.  I think that's a good idea too!

File: lang/py/avro/io.py (hunk 1 above)
Comment (jjaakola-aiven) [reply lang/py/avro/io.py:728]: The proposed idea here is to map the specific read error to something that helps out when encoding is off. Logging the data and schema used would help finding the issue.
But I'll remove this. I think logging the data is wrong, as there are environments where data in the datum is sensitive.

File: lang/py/avro/errors.py [hunk 3]
```diff
@@ -32,6 +32,10 @@ class AvroException(Exception):
     """The base class for exceptions in avro."""
 
 
+class InvalidBytesRead(AvroException):
```
[the remark this answers is not in the document]
Comment (jjaakola-aiven) [reply lang/py/avro/errors.py]: Renamed to `InvalidAvroBinaryEncoding`.

File: lang/py/avro/io.py (hunk 2 above)
Comment (jjaakola-aiven) [reply lang/py/avro/io.py:227]: Added, raises `InvalidAvroBinaryEncoding`.

**Pull request merged**